### PR TITLE
STorM32 native serial mount

### PR DESCRIPTION
**Pull request by olliw42**
The PR is about supporting STorM32 over serial UART.

It is motivated by a statement in a recent dev call, "So long as no harm is done outside we should accept new functionality" (https://discuss.ardupilot.org/t/dev-call-apr-23-2018-2300-utc/28358/2). The PR is as isolated as it could be, and does not affect any outside, IMHO. It very closely follows AP_Mount_SToRM32_serial, and only infects the AP_SerialManager.cpp/.h and AP_Mount.cpp/.h library files in trivial ways.

The PR doesn't provide the best possible STorM32 support, but should represent an improvement over alternatives:
 - proper targeting in all mount modes, Solo smart shots, gimbal paddle etc. working
 - STorM32 functions and scripts
 - STorM32-Link horizon and yaw drift compensation

It has been tested to compile for ArduCopter with px4-v2 (dec = 1085520), px4-v3, px4-v4, has been bench tested for px4-v4 using a pixracer, and flight tested for px4-v3 using a Solo (log and video of a test-flight: www.olliw.eu/drop/ap/pr-storm32nativeonly-logs.zip, www.olliw.eu/drop/ap/pr-storm32nativeonly-video.wmv).

### CODE REVIEW

**Comment by Pedals2Paddles** [comment]
Which Storm32 controllers can utilize this?

**Comment by olliw42** [comment]
any
but I think you rather wanted to know which setups/firmwares can use this: STorM32 NT and T-STorM32

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+private:
+    // helper to handle corrupted rcin data
+    bool is_failsafe(void);
+
+    // interface to STorM32_lib
+    virtual size_t _serial_txspace(void);
+    virtual size_t _serial_write(const uint8_t *buffer, size_t size, uint8_t priority);
+    virtual uint32_t _serial_available(void);
+    virtual int16_t _serial_read(void);
+    uint16_t _rcin_read(uint8_t ch);
```
Comment (OXINARF) [inline libraries/AP_Mount/AP_Mount_STorM32_native.h]: Same about `override`.

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    } _status;
+    bool _status_updated;
+
+    void set_status_angles_deg(float pitch_deg, float roll_deg, float yaw_deg);
+    void get_status_angles_deg(float* pitch_deg, float* roll_deg, float* yaw_deg);
+
+    // target out
+    enum ANGLESTYPEENUM {
+        angles_deg = 0, //the STorM32 convention is angles in deg, not rad!
+        angles_pwm
```
Comment (OXINARF) [inline libraries/AP_Mount/AP_Mount_STorM32_native.h]: Can these enum values be changed to uppercase? They look like variables in code instead of enum values.

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    };
+
+    struct {
+        enum MAV_MOUNT_MODE mode;
```
Comment (OXINARF) [inline libraries/AP_Mount/AP_Mount_STorM32_native.h]: Why is this needed? You can get the mount mode at any time with `get_mode()`.

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+//******************************************************
+// STorM32_lib class functions
+//******************************************************
+
+/// Constructor
+STorM32_lib::STorM32_lib() :
+    _serial_is_initialised(false),
+    _storm32link_seq(0)
+{
+    _serial_in.state = SERIALSTATE_IDLE;
```
Comment (OXINARF) [inline libraries/AP_Mount/STorM32_lib.cpp]: Same comments: zero initialization isn't needed and this one can be put in class declaration.

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+}
+
+// determines from the STorM32 state if the gimbal is in a normal operation mode
+bool STorM32_lib::is_normal_state(uint16_t state)
+{
+    if ((state == STORM32STATE_NORMAL) || (state == STORM32STATE_STARTUP_FASTLEVEL)) { return true; }
```
Comment (OXINARF) [inline libraries/AP_Mount/STorM32_lib.cpp]: Style

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+    t.dummy3 = 0;
+    t.dummy4 = 0;
+    t.dummy5 = 0;
+    t.crc = crc_calculate(&(t.len), sizeof(tCmdDoCamera)-3);
+
+    _serial_write( (uint8_t*)(&t), sizeof(tCmdDoCamera) );
+}
+
+// 28 bytes = 2431us @ 115200bps
+void STorM32_lib::send_cmd_setinputs(void)
```
Comment (OXINARF) [inline libraries/AP_Mount/STorM32_lib.cpp]: Unused.

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+    t.channel14 = _rcin_read(14);
+    t.channel15 = _rcin_read(15);
+    t.status = status;
+    t.crc = crc_calculate(&(t.len), sizeof(tCmdSetInputs)-3);
+
+    _serial_write( (uint8_t*)(&t), sizeof(tCmdSetInputs) );
+}
+
+// 19 bytes = 1650us @ 115200bps
+void STorM32_lib::send_cmd_sethomelocation(const AP_AHRS_TYPE &ahrs)
```
Comment (OXINARF) [inline libraries/AP_Mount/STorM32_lib.cpp]: Unused.

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+    t.longitude = location.lng;
+    t.altitude = location.alt;
+    t.status = status;
+    t.crc = crc_calculate(&(t.len), sizeof(tCmdSetHomeTargetLocation)-3);
+
+    _serial_write( (uint8_t*)(&t), sizeof(tCmdSetHomeTargetLocation) );
+}
+
+// 19 bytes = 1650us @ 115200bps
+void STorM32_lib::send_cmd_settargetlocation(void)
```
Comment (OXINARF) [inline libraries/AP_Mount/STorM32_lib.cpp]: Unused

File: libraries/AP_Mount/STorM32_lib.h [hunk 9]
```diff
@@ -0,0 +1,255 @@
+#pragma once
+
+#include <AP_AHRS/AP_AHRS.h>
+
+#define SERIAL_RECEIVE_BUFFER_SIZE      96 //the largest RCcmd response can be 77
```
Comment (OXINARF) [inline libraries/AP_Mount/STorM32_lib.h]: Make this more unique, something like `STORM32_LIB_SERIAL_RECEIVE_BUFFER_SIZE`

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+//------------------------------------------------------
+
+// init - performs any required initialisation for this instance
+void AP_Mount_STorM32_native::init(const AP_SerialManager& serial_manager)
+{
+    //from instance we can determine its type, we keep it here since that's easier/shorter
+    _mount_type = _frontend.get_mount_type(_instance);
+
+    // it should never happen that it's not one of them, but let's enforce it, to not depend on the outside
+    if (_mount_type != AP_Mount::Mount_Type_STorM32_native) {
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: _mount_type serves two purposes:
* the same code can be used for various serial data streams, with little modifications
examples would be UART and CAN
* I "missuse" _mount_type as a state variable. This can be avoided, but would need adding then an additional state variable ... I found that a bit too much, so I went this way

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+        _mount_type = AP_Mount::Mount_Type_None;
+    }
+
+    if (_mount_type == AP_Mount::Mount_Type_STorM32_native) {
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: same as before

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+        _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_STorM32_Native, 0);
+        if (_uart) {
+            _serial_is_initialised = true; //tell the STorM32_lib class
+            //we do not set _initialised = true, since we first need to pass find_gimbal()
+        } else {
+            _serial_is_initialised = false; //tell the STorM32_lib class
+            _mount_type = AP_Mount::Mount_Type_None; //this prevents many things from happening, safety guard
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: see explanation in main post

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+    }
+
+    //slow down everything to 100 Hz
+    // we can't use update(), since 50 Hz isn't compatible with the desired 20 Hz STorM32Link rate
+    // each message is send at 20 Hz i.e. 50 ms, for 5 task slots => 10 ms per task slot
+    uint64_t current_time_ms = AP_HAL::millis64();
+    if ((current_time_ms - _task_time_last) >= 10) {
+        _task_time_last = current_time_ms;
+
+        const uint16_t LIVEDATA_FLAGS = LIVEDATA_STATUS_V2|LIVEDATA_ATTITUDE_RELATIVE;
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: ah, I missed that style guide, sorry

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+        switch (_task_counter) {
+            case TASK_SLOT0:
+                if (_bitmask & SEND_STORM32LINK_V2) {
+                    send_storm32link_v2(_frontend._ahrs); //2.3ms
+                }
+
+                break;
+            case TASK_SLOT1:
+                // trigger live data
+                if (_mount_type == AP_Mount::Mount_Type_STorM32_native) {
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: it will allow further PRs to come :)

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+                break;
+            case TASK_SLOT3:
+                if (_bitmask & SEND_CMD_SETINPUTS) {
+                    send_cmd_setinputs(); //2.4ms
+                }
+
+                break;
+            case TASK_SLOT4:
+                // receive live data
+                if (_mount_type == AP_Mount::Mount_Type_STorM32_native) {
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: the code can be used with different serial streams, but this one here is only allowed for UART

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+    enum MAV_MOUNT_MODE mount_mode = get_mode();
+
+    switch (mount_mode) {
+        // move mount to a "retracted" position.
+        case MAV_MOUNT_MODE_RETRACT:
+            {
+                const Vector3f &target = _state._retract_angles.get();
+                _angle_ef_target_rad.x = ToRad(target.x);
+                _angle_ef_target_rad.y = ToRad(target.y);
+                _angle_ef_target_rad.z = ToRad(target.z);
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: ah, ok ...

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+            if (get_pwm_target_from_radio) {
+                get_pwm_target_angles_from_radio(&pitch_pwm, &roll_pwm, &yaw_pwm);
+                send_pwm_target = true;
+            } else {
+                update_targets_from_rc();
+                send_ef_target = true;
+            }
+            if (is_failsafe()) {
+                pitch_pwm = roll_pwm = yaw_pwm = 1500;
+                _angle_ef_target_rad.y = _angle_ef_target_rad.x = _angle_ef_target_rad.z = 0.0f;
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: Pl note that the STorM32 has its own neutral position, and I've made the code using this (camera recenter)

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 18]
```diff
@@ -0,0 +1,497 @@
...
+            // we do not know this mode so do nothing
+            break;
+    }
+
+    // send target angles
+    if (send_ef_target) {
+        set_target_angles_rad(_angle_ef_target_rad.y, _angle_ef_target_rad.x, _angle_ef_target_rad.z, mount_mode);
+    }
+
+    if (send_pwm_target) {
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: no, they can't, as you say

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 19]
```diff
@@ -0,0 +1,497 @@
...
+    #define rc_ch(i) RC_Channels::rc_channel(i-1)
+
+    if (rc_in && (rc_ch(rc_in))) {
+        *pwm = rc_ch(rc_in)->get_radio_in();
+    } else {
+        *pwm = 1500;
+    }
+}
+
+void AP_Mount_STorM32_native::set_target_angles_deg(float pitch_deg, float roll_deg, float yaw_deg, enum MAV_MOUNT_MODE mount_mode)
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: ähm, that's strange, I thought I used it ... I'll double check what's up

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 20]
```diff
@@ -0,0 +1,497 @@
...
+    _target.mode = mount_mode;
+    _target_to_send = true;
+}
+
+void AP_Mount_STorM32_native::send_target_angles(void)
+{
+    if (_target.mode <= MAV_MOUNT_MODE_NEUTRAL) { //RETRACT and NEUTRAL
+        if (_target_mode_last != _target.mode) { // only do it once, i.e., when mode has just changed
+            // trigger a recenter camera, this clears all internal Remote states
+            // the camera does not need to be recentered explicitly, thus return
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: Pl note that the STorM32 has its own neutral position, and I've made the code using this (camera recenter)

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 21]
```diff
@@ -0,0 +1,497 @@
...
+    if (!_serial_is_initialised) {
+        return;
+    }
+
+    uint64_t current_time_ms = AP_HAL::millis64();
+
+#if FIND_GIMBAL_MAX_SEARCH_TIME_MS
+    if (current_time_ms > FIND_GIMBAL_MAX_SEARCH_TIME_MS) {
+        _initialised = false; //should be already false, but it can't hurt to ensure that
+       _serial_is_initialised = false; //switch off BP_STorM32
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: oh, I missed to adapt this to this PR :)

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 22]
```diff
@@ -0,0 +1,497 @@
...
+                    versionstr[16] = '\0';
+                    for (uint16_t n=0;n<16;n++) boardstr[n] = _serial_in.getversionstr.boardstr[n];
+                    boardstr[16] = '\0';
+                    _task_counter = TASK_SLOT0;
+                    _initialised = true;
+                }
+                break;
+        }
+        _task_counter++;
+        if (_task_counter >= 3) { _task_counter = 0; }
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: this is for timing
it's easier than adding some additional stuff just to do some delay

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 23]
```diff
@@ -0,0 +1,497 @@
...
+    }
+}
+
+//------------------------------------------------------
+// interfaces to STorM32_lib
+//------------------------------------------------------
+
+size_t AP_Mount_STorM32_native::_serial_txspace(void)
+{
+    if (_mount_type == AP_Mount::Mount_Type_STorM32_native) {
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: pl see respective comment in main post

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 24]
```diff
@@ -0,0 +1,497 @@
...
+        return _uart->available();
+    }
+    return 0;
+}
+
+
+int16_t AP_Mount_STorM32_native::_serial_read(void)
+{
+    if (_mount_type == AP_Mount::Mount_Type_STorM32_native) {
+        return _uart->read();
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: same

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 25]
```diff
@@ -0,0 +1,497 @@
...
+    }
+    return 0;
+}
+
+
+uint16_t AP_Mount_STorM32_native::_rcin_read(uint8_t ch)
+{
+    return hal.rcin->read(ch);
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: I'll double check what RC_Channels is offereing
but I think: that's the whole point not to use RC_Channels here

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 26]
```diff
@@ -0,0 +1,138 @@
...
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <AP_GPS/AP_GPS.h>
+#include <GCS_MAVLink/GCS.h>
+#include <RC_Channel/RC_Channel.h>
+#include "AP_Mount.h"
+#include "AP_Mount_Backend.h"
+#include "STorM32_lib.h"
+
+#define FIND_GIMBAL_MAX_SEARCH_TIME_MS  90000 //AP's startup has become quite slow, so give it plenty of time, set to 0 to disable
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: I guess the lib will be then more stylish than many other libs :D

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+
+class AP_Mount_STorM32_native : public AP_Mount_Backend, public STorM32_lib
+{
+
+public:
+    // Constructor
+    AP_Mount_STorM32_native(AP_Mount &frontend, AP_Mount::mount_state &state, uint8_t instance);
+
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: ok

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+    virtual void update_fast();
+
+    // has_pan_control - returns true if this mount can control it's pan (required for multicopters)
+    virtual bool has_pan_control() const { return false; }
+
+    // set_mode - sets mount's mode
+    virtual void set_mode(enum MAV_MOUNT_MODE mode);
+
+    // status_msg - called to allow mounts to send their status to GCS using the MOUNT_STATUS message
+    virtual void status_msg(mavlink_channel_t chan);
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: you hit my limited C++ knowledge
I always though that's what virtual is

could you please tell me for one example how exactly it should look like?

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+
+    // every mount should have this !!!
+    virtual bool is_armed(){ return _armed; }
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: ähm, I thought I use that, I'll double check

File: libraries/AP_Mount/AP_Mount_STorM32_native.h (hunk 1 above)
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: same

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+        TASK_SLOTNUMBER,
+    };
+    uint64_t _task_time_last;
+    uint16_t _task_counter;
+
+    // we want to keep that info, the fields are 16 in size, so add one to convert it to string
+    //  _initialised = true indicates that these fields were set
+    char versionstr[16+1];
+    char namestr[16+1];
+    char boardstr[16+1];
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: this will allow follow up PRs

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 31]
```diff
@@ -0,0 +1,138 @@
...
+    };
+    uint16_t _bitmask; //this mask is to control some functions
+
+    // storm32.Status in
+    struct {
+        float pitch_deg;
+        float roll_deg;
+        float yaw_deg;
+    } _status;
+    bool _status_updated;
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: really? I'll double check

File: libraries/AP_Mount/AP_Mount_STorM32_native.h (hunk 3 above)
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: pl note that there is a stateness in how I use mode for handling the camera recenter
I alternatively could add an extra variable for that, but one is needed

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 32]
```diff
@@ -0,0 +1,404 @@
...
+    Vector3f vel;
+    //ahrs.get_velocity_NED(vel) returns a bool, what does it exactly mean ???
+    // whatever it means it's probably a good idea to consider it
+    if (!ahrs.get_velocity_NED(vel)) { vel.x = vel.y = vel.z = 0.0f; }
+
+    tSTorM32LinkV2 t;
+    t.stx = 0xF9;
+    t.len = 0x21;
+    t.cmd = 0xDA;
+    t.seq = _storm32link_seq; _storm32link_seq++;
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.cpp]: I know, but any compiler should handle that. And I find that much more readable.
Pl note my comment as regards STorM32_lib in the main post

File: libraries/AP_Mount/STorM32_lib.cpp (hunk 6 above)
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.cpp]: see comment in main post

File: libraries/AP_Mount/STorM32_lib.cpp (hunk 7 above)
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.cpp]: see comment in main post

File: libraries/AP_Mount/STorM32_lib.cpp (hunk 8 above)
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.cpp]: dito

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 33]
```diff
@@ -0,0 +1,404 @@
...
+//------------------------------------------------------
+
+void STorM32_lib::receive_reset(void)
+{
+    _serial_in.state = SERIALSTATE_IDLE;
+}
+
+void STorM32_lib::receive_reset_wflush(void)
+{
+    while (_serial_available() > 0) {
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.cpp]: ???
I don't understand
any available function in all other code bases I know allow to check how many bytes are there without crashing
what's so special/different about AP's available()?

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 34]
```diff
@@ -0,0 +1,404 @@
...
+// call a flush_rx() and receive_reset() to take care of both of that
+void STorM32_lib::do_receive_singlechar(void)
+{
+    uint8_t c;
+
+    if (_serial_available() <= 0) { return; } //this should never happen, but play it safe
+
+    switch (_serial_in.state) {
+        case SERIALSTATE_IDLE:
+            c = _serial_read();
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.cpp]: that's correct

File: libraries/AP_Mount/STorM32_lib.h (hunk 9 above)
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.h]: ok

File: libraries/AP_Mount/STorM32_lib.h [hunk 35]
```diff
@@ -0,0 +1,255 @@
...
+        SERIALSTATE_RECEIVE_PAYLOAD_LEN,
+        SERIALSTATE_RECEIVE_CMD,
+        SERIALSTATE_RECEIVE_PAYLOAD,
+        SERIALSTATE_MESSAGE_RECEIVED,
+        SERIALSTATE_MESSAGE_RECEIVEDANDDIGESTED,
+    };
+
+    typedef struct { //structure to process incoming serial data
+        // auxiliary fields to handle reception
+        uint16_t state;
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/STorM32_lib.h]: yes

**Comment by olliw42** [comment]
>So, STorM32 supports three protocols now (MAVLink, custom serial and native serial), is that correct?

not exactly, all these protocols are there since 4 years or so (the command set has seen some variation though), i.e. it's not like there would be suddenly an all new protocol ... the mount did need a new name, and I chose "native".

>I've made some comments, mostly about style. The use of _mount_type, _serial_is_initialised and _uart != >nullptr is quite mixed - just using the last one seems perfectly sufficient.

I don't think so. Please note that the STorM32_lib class is designed such to handle any serial data stream, especially also CAN (if you would want to see how it looks with CAN you could look into the betacopter branch).

The STorM32_lib library I guess could/should be considered similarly to other 3rd party libs (Mavlink, blabla). That is , I think it should be acceptable that it has a fuller support of the STorM32 protocol than is actually used by AP. If you prefer I make it a git submodule.

As regards the initialization, I noticed that there are different approaches in the different code parts, and I also noticed that what was preferred has changed several times over time. That's why, and because I have now knowledge myself, why it is admittedly quite mixed. Could you please tell by means of an example what your current preference is, e.g. a line reference to another code file?
(I can make a guess what "initialization list", "class declaration", etc. pp. could mean, but I don't really know).
Your comment seems to imply that only those fields which are not zero need to be set, correct?

**Comment by olliw42** [comment]
dammed, I again pressed the wrong button ...

File: libraries/AP_Mount/AP_Mount_STorM32_native.h [hunk 36]
```diff
@@ -0,0 +1,138 @@
...
+                float yaw;
+            } deg;
+            struct {
+                uint16_t pitch;
+                uint16_t roll;
+                uint16_t yaw;
+            } pwm;
+        };
+    } _target;
+    bool _target_to_send;
```
[the remark this answers is not in the document]
Comment (olliw42) [reply libraries/AP_Mount/AP_Mount_STorM32_native.h]: EDIT: this is correct, but it would be need for CAN ... I guess you want me to remove nevertheless, so I remove

**Comment by olliw42** [comment]
EDIT: ah, now I see what you meant with that _uart != nullptr thing ... yes, you are right, this one is wrong, good catch

**Comment by olliw42** [comment]
ah, I just did the changes as far as I could and pushed them to my fork branch ...

now I see your changes

not sure what I should do now
?

**Comment by Pedals2Paddles** [comment]
When you push changes to the branch this PR represents, it will automatically be reflected in the PR. That goes for adding commits, or force-pushing a complete redo.

**Comment by olliw42** [comment]
thx
in the meantime I also have discovered that small line below the comments saying "Add more commits by pushing to the storm32nativeonly branch on olliw42/ardupilot." ... finding the text and reading helps :D

**Comment by olliw42** [comment]
IMHO I have made all suggested corrections or gave an answer, from my side I thus wouldn't know what further to do at this point. just for the records :)

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 37]
```diff
@@ -0,0 +1,508 @@
...
+{
+    if (!_initialised) {
+        return;
+    }
+
+    //slow down everything to 100 Hz
+    // we can't use update(), since 50 Hz isn't compatible with the desired 20 Hz STorM32Link rate
+    // each message is send at 20 Hz i.e. 50 ms, for 5 task slots => 10 ms per task slot
+    uint64_t current_time_ms = AP_HAL::millis64();
+    if ((current_time_ms - _task_time_last) >= 10) {
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: why not use 32 bit millis and unsigned subtraction?

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 38]
```diff
@@ -0,0 +1,508 @@
...
+int16_t AP_Mount_STorM32_native::_serial_read(void)
+{
+    if (_mount_type == AP_Mount::Mount_Type_STorM32_native) {
+        return _uart->read();
+    }
+    return 0;
+}
+
+uint16_t AP_Mount_STorM32_native::_rcin_read(uint8_t ch)
+{
```
Comment (tridge) [inline libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: unused function?

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp [hunk 39]
```diff
@@ -1,26 +1,20 @@
...
-    _task_time_last = 0;
-    _task_counter = TASK_SLOT0;
-
     _bitmask = SEND_STORM32LINK_V2 | SEND_CMD_SETINPUTS | SEND_CMD_DOCAMERA;
-
-    _status.pitch_deg = _status.roll_deg = _status.yaw_deg = 0.0f;
-    _status_updated = false;
-
-    _target_to_send = false;
-    _target_mode_last = MAV_MOUNT_MODE_RETRACT;
```
Comment (OXINARF) [inline libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: Removing this last one seems wrong as I don't see it added to class declaration.

File: libraries/AP_Mount/STorM32_lib.cpp [hunk 40]
```diff
@@ -334,7 +338,11 @@ void STorM32_lib::receive_reset(void)
 
 void STorM32_lib::receive_reset_wflush(void)
 {
-    while (_serial_available() > 0) {
+    //it is claimed a while can't do it, so do a for
+    for (uint32_t i = 0; i < _serial_available(); i++) {
```
Comment (OXINARF) [inline libraries/AP_Mount/STorM32_lib.cpp]: Why do you think this is OK? `_serial_available` will be called in every loop and that's the issue, it might take a long time to be less or equal than `i`.

File: libraries/AP_Mount/AP_Mount_STorM32_native.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/AP_Mount/AP_Mount_STorM32_native.cpp]: > the same code can be used for various serial data streams, with little modifications
examples would be UART and CAN

Ok, that's a reasonable use.

> I "missuse" _mount_type as a state variable. This can be avoided, but would need adding then an additional state variable ... I found that a bit too much, so I went this way

I don't see why you need an additional state variable. And you definitely shouldn't use this as a state variable.

File: libraries/AP_Mount/STorM32_lib.cpp (hunk 33 above)
[the remark this answers is not in the document]
Comment (OXINARF) [reply libraries/AP_Mount/STorM32_lib.cpp]: You can check how many bytes are available, you can't do it continuously because this will be run in the main flight thread.

**Comment by olliw42** [comment]
many thx for reviewing
based on last week's devcall I however retract this PR.

I definitely do not want to pollute your high-quality code with low-quality code. Since it is as you say I better close this PR.

I also do not want to be responsible for code with which I wouldn't agree in all points.

I seriously have no idea why one would want to go through all this in one's precious spare time, I couldn't find any enjoyable moment in it, to the contrary. All my fault, my apologies.
Wish you all the best, OW

**Comment by Evan1010** [comment]
@OXINARF Out of curiosity, what is the future of such a closed PR? Can others still work on it and eventually reopen it?

I'm a little confused on which reviewer comments are still under contention, but your latest review comments seem to be issues I can understand and address. I'll keep going through the prior reviewer comments in the meantime.

I'd be saddened to see this go away permanently, especially when it was demonstrated as functional, with method and consistency issues here and there.

**Comment by khancyr** [comment]
@Evan1010 you can take on this PR !

**Comment by OXINARF** [comment]
@Evan1010 I will (hopefully) be working on it this week, as I said in the dev call.

**Comment by Evan1010** [comment]
@OXINARF Good to hear.
